### PR TITLE
token-2022: Add confidential transfer with fee instruction

### PR DESCRIPTION
**Pull request by samkim-crypto**
I added a separate instruction for confidential transfer with fee. 

Confidential transfer with and without fees require different zkps. I was originally considering having a single confidential transfer instruction for either with or without fees. The processor would decode the zk verify instruction and distinguish the fee/no fee case. But since the two zkps require different inputs, I think the client code will be a bit simpler to have two separate instructions.

### CODE REVIEW

**Comment by samkim-crypto** [comment]
@mvines Can you look at this briefly to check if this looks reasonable? 

The part that I am unsure about is the new instruction `TransferWithFee` and the `confidential_transfer_transfer_with_fee` in `client` that takes in an epoch information.

The changes in `program-2022-test` are going to be updated with decryption once zk-token-sdk 0.8.0 becomes available, so you can probably take a more detailed look in a future PR.

**Comment by samkim-crypto** [comment]
Hm, for some reason, the test-bpf passes on my local machine, but it seems to fail here. I must done some commit wrong. I will figure it out soon...

**Comment by samkim-crypto** [comment]
Hm... I seems to always get EqualityProof error, but all the proofs seem to verify on the client side https://github.com/solana-labs/solana-program-library/pull/2988/commits/734d3f7a92efc77d435f1e0a73561642f486a371. Not sure what is going on...

**Comment by mvines** [comment]
> Hm... I seems to always get EqualityProof error, but all the proofs seem to verify on the client side [734d3f7](https://github.com/solana-labs/solana-program-library/commit/734d3f7a92efc77d435f1e0a73561642f486a371). Not sure what is going on...

You could try instrumenting the code around https://github.com/solana-labs/solana/blob/949006b5a2559b57dd5967963b44a232277f63e2/programs/zk-token-proof/src/lib.rs#L62 to see if something different is getting passed into the actual proof program perhaps?

File: token/client/src/token.rs [hunk 1]
```diff
@@ -1148,6 +1150,95 @@ where
...
+        )?;
+
+        let mint_state = self.get_mint_info().await.unwrap();
+        let transfer_fee_config = mint_state
+            .get_extension::<transfer_fee::TransferFeeConfig>()
+            .unwrap();
+
+        // Zkp generator requires transfer fee parameters as input
+        let fee_parameters =
+            if u64::from(transfer_fee_config.newer_transfer_fee.epoch) < epoch_info.epoch {
```
Comment (mvines) [inline token/client/src/token.rs:1185]: It looks like you could use `transfer_fee_config.get_epoch_fee(epoch_info.epoch)` to obtain the appropriate `TransferFee`, which'll save a couple lines of code here

Comment (samkim-crypto) [reply token/client/src/token.rs]: Yep this is applied now, thanks!

**Pull request merged**